models/issue_dependency.go

```diff
@@ -0,0 +1,169 @@
+// Copyright 2018 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package models
+
+import (
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+	"code.gitea.io/gitea/modules/util"
+)
+
+// IssueDependency represents an issue dependency
+type IssueDependency struct {
+	ID           int64          `xorm:"pk autoincr"`
+	UserID       int64          `xorm:"NOT NULL"`
+	IssueID      int64          `xorm:"UNIQUE(issue_dependency) NOT NULL"`
+	DependencyID int64          `xorm:"UNIQUE(issue_dependency) NOT NULL"`
+	CreatedUnix  util.TimeStamp `xorm:"INDEX created"`
+	UpdatedUnix  util.TimeStamp `xorm:"updated"`
+}
+
+// DependencyType Defines Dependency Type Constants
+type DependencyType int
+
+// Define Dependency Types
+const (
+	DependencyTypeBlockedBy DependencyType = iota
+	DependencyTypeBlocking
+)
+
+// CreateIssueDependency creates a new dependency for an issue
+func CreateIssueDependency(user *User, issue, dep *Issue) (err error) {
+	sess := x.NewSession()
+	defer sess.Close()
+	if err = sess.Begin(); err != nil {
+		return err
+	}
+
+	// Check if it aleready exists
+	exists, err := issueDepExists(sess, issue.ID, dep.ID)
+	if err != nil {
+		return err
+	}
+	if exists {
+		return ErrDependencyExists{issue.ID, dep.ID}
+	}
+	// And if it would be circular
+	circular, err := issueDepIsCircular(sess, issue.ID, dep.ID)
+	if err != nil {
+		return err
+	}
+	if circular {
+		return ErrCircularDependency{issue.ID, dep.ID}
+	}
+
+	// No error means the dependency doesn't exist nor is circular, so we can proceed to create a new one.
+	newIssueDependency := &IssueDependency{
+		UserID:       user.ID,
+		IssueID:      issue.ID,
+		DependencyID: dep.ID,
+	}
+
+	if _, err := sess.Insert(newIssueDependency); err != nil {
+		return err
+	}
+
+	// Add comment referencing the new dependency
+	if _, err = createIssueDependencyComment(sess, user, issue, dep, true); err != nil {
```

Comment:
If you always call `createIssueDependencyComment` twice (same in `RemoveIssueDependency`), is not better to change it so that function will add both comments at once?

Reply:
Yes, I fixed that as well.